Document versioning of CRI API (#114394)
kubernetes/kubernetes: staging/src/k8s.io/cri-api/README.md

version'd kubelet and older versioned crictl. This is a supported scenario within
the version skew policy.

### Version Skew Policy for CRI API

CRI API has two versions:
- Major semantic version (known versions are `v1alpha2` ([removed in 1.26](https://kubernetes.io/blog/2022/12/09/kubernetes-v1-26-release/#cri-v1alpha2-removed)), `v1`).
- Kubernetes version (for example: `@1.23`). Note, the `cri-api` Golang library is versioned as `0.23` as it doesn't guarantee Go types backward compatibility.

Major semantic version (e.g. `v1`) is used to introduce breaking changes
and major new features that are incompatible with the current API.

Kubernetes version is used to indicate a specific feature set implemented
on top of the major semantic version. All changes made without the change
of a major semantic version API must be backward and forward compatible.

- _Kubelet_ must work with the older _Container Runtime_ if it implements
  the same semantic version of CRI API (e.g. `v1`) of up to three Kubernetes minor
  versions back. New features implemented in CRI API must be gracefully degraded.
  For example, _Kubelet_ of version 1.26 must work with _Container Runtime_

>   implementing `k8s.io/cri-api@v0.23.0`+.

Reviewer: I think we must flesh out what it means in terms of proto changes that are allowed to support any skew. Also, examples of how things will behave when kubelet and runtime aren't on the same version will be helpful to make a call here.

Reviewer 2: Thanks @SergeyKanzhelev, this document is really much appreciated 👍

I too would like to understand a bit better what kind of breaking changes would be allowed inside a stable major semantic version (`v1`). But OTOH I think we can merge this in it's current form and making further improvements in separate PRs ->
/lgtm

on backward compatibility of Go wrappers between versions. However CRI API itself
(protocol buffers and gRPC API) is marked as stable `v1` version and it is
backward compatible between versions.

Versions like `v0.<minor>.<patch>` (e.g. `v0.25.5`) are considered stable.
It is discouraged to introduce CRI API changes in patch releases and recommended
to use versions like `v0.<minor>.0`.

All alpha and beta versions (e.g. `k8s.io/cri-api@v0.26.0-beta.0`) should be
backward and forward compatible.

## Feature development

Some features development requires changes in CRI API and corresponding changes
in _Container Runtime_. Coordinating between Kubernetes branches and release
versions and _Container Runtime_ versions is not always trivial.

The recommended feature development flow is following:

- Review proposed CRI API changes during the KEP review stage.
  Some field names and types may not be spelled out exactly at this stage.
- Locally implement a prototype that implement changes in both - Kubernetes and Container Runtime.
- Submit a Pull Request for Kubernetes implementing CRI API changes alongside the feature code.

>   Feature must be developed to degrade gracefully when used with older Container Runtime

Reviewer: Example of what this means in practice on both the kubelet and runtime side will be useful here.

> Please do NOT file issues or submit PRs against the [kubernetes/cri-api](https://github.com/kubernetes/cri-api)

Reviewer 3: A matching emphasis change
```suggestion
Please do **not** file issues or submit PRs against the [kubernetes/cri-api](https://github.com/kubernetes/cri-api)
```

repository as it is readonly, all development is done in [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes).

> CRI API is defined in [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes)
> repository and is ONLY intended to be used for kubelet to container runtime interactions.
> It is NOT a common purpose container runtime API for general use and intended
> to be Kubernetes-centric. We try to avoid it, but there may be logic in container
> runtime optimized for the order or specific parameters of call kubelet makes.

Reviewer 3: - Markdown has native support for strong emphasis.
- `crictl` is a thing

```suggestion
The CRI API is defined in [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes)
repository and is **only** intended to be used for kubelet to container runtime 
interactions, or for node-level troubleshooting using a tool such as `crictl`.
It is **not** a common purpose container runtime API for general use, and is intended
to be Kubernetes-centric. We try to avoid it, but there may be logic within a container
runtime that optimizes for the order or specific parameters of call(s) that the kubelet
makes.
```

different versions of CRI API.

For example, on a single node there might be:

- _Kubelet_ may call into _Container Runtime_ (e.g. [containerd](https://containerd.io))
  and _Image Service Proxy_ (e.g. [stargz-snapshotter](https://github.com/containerd/stargz-snapshotter)).
  _Container Runtime_ may be versioned with the OS Image, _Kubelet_ is installed
  by system administrator and _Image Service proxy_ is versioned by the third party vendor.
- _Image Service Proxy_ calls into _Container Runtime_.
- _CRI tools_ (e.g. [crictl](https://kubernetes.io/docs/tasks/debug/debug-cluster/crictl/))
  may be installed by end user to troubleshoot, same as a third party daemonsets.
  All of them are used to call into the _Container Runtime_ to collect container information.

So on a single node it may happen that _Container Runtime_ is serving a newer
version'd kubelet and older versioned crictl. This is a supported scenario within
the version skew policy.

### Version Skew Policy for CRI API

CRI API has two versions:
- Major semantic version (known versions are `v1alpha2` ([removed in 1.26](https://kubernetes.io/blog/2022/12/09/kubernetes-v1-26-release/#cri-v1alpha2-removed)), `v1`).

> - Kubernetes version (`@1.23`). Note, the library cri-api is versioned as `0.23` as it doesn't guarantee Go types backward compatibility.

Reviewer 3: ```suggestion
- Kubernetes version (for example: `@1.23`). Note, the `cri-api` Golang library is versioned as `0.23` as it doesn't guarantee Go types backward compatibility.
```

backward and forward compatible.

## Feature development

Some features development requires changes in CRI API and corresponding changes
in _Container Runtime_. Coordinating between Kubernetes branches and release
versions and _Container Runtime_ versions is not always trivial.

The recommended feature development flow is following:

- Review proposed CRI API changes during the KEP review stage.
  Some field names and types may not be spelled out exactly at this stage.
- Locally implement a prototype that implement changes in both - Kubernetes and Container Runtime.
- Submit a Pull Request for Kubernetes implementing CRI API changes alongside the feature code.
  Feature must be developed to degrade gracefully when used with older Container Runtime
  according to the Version Skew policy.
- Once PR is merged, wait for the next Kubernetes release tag being produced.
  Find the corresponding CRI API tag (e.g. `k8s.io/cri-api@v0.26.0-beta.0`).
- This tag can be used to implement the feature in Container Runtime. It is recommended
  to switch to the stable tag like (`k8s.io/cri-api@v0.26.0`) once available.

## Change history

> Here is the change history of CRI API.

Reviewer 3: Which is right:
```suggestion
Here is the change history of `cri-api`:
```
or
```suggestion
Here is the change history of the Container Runtime Interface protocol:
```
?

I couldn't work it out.
